pkg/crc/cluster/cert-renewal.go

```diff
@@ -1,54 +1,72 @@
 			return nil
 		}
-		return &errors.RetriableError{Err: fmt.Errorf("No Pending CSR with signerName %s", signerName)}
-	}
-
-	return errors.RetryAfter(8*time.Minute, waitForPendingCsr, time.Second*5)
+		return &crcerrors.RetriableError{Err: fmt.Errorf("No Pending CSR with signerName %s", signerName)}
+	}, time.Second*5)
 }
 
-func WaitAndApprovePendingCSRs(ocConfig oc.Config, client, server bool) error {
-	/* 2 CSRs to approve, one right after kubelet restart, the other one a few dozen seconds after
-	approving the first one
-	- First one is requested by system:serviceaccount:openshift-machine-config-operator:node-bootstrapper
-	- Second one is requested by system:node:<node_name> */
+func ApproveCSRAndWaitForCertsRenewal(sshRunner *ssh.Runner, ocConfig oc.Config, client, server bool) error {
+	// First, kubelet starts and tries to connect to API server. If its certificate is expired, it asks for a new one
+	// Admin needs to approve it. The Kubernetes controller manager will then issue the cert, kubelet will fetch it and use it.
+	// Kubelet stores the cert in /var/lib/kubelet/pki/kubelet-client-current.pem
 	if client {
-		if err := waitForPendingCsrs(ocConfig, "kubernetes.io/kube-apiserver-client-kubelet"); err != nil {
+		if err := waitForPendingCsrs(ocConfig, kubeletClientSignerName); err != nil {
 			logging.Debugf("Error waiting for first pending (node-bootstrapper) CSR: %v", err)
 			return err
 		}
-		if err := ApproveNodeCSR(ocConfig, "kubernetes.io/kube-apiserver-client-kubelet"); err != nil {
+		if err := approveNodeCSR(ocConfig, kubeletClientSignerName); err != nil {
 			logging.Debugf("Error approving first pending (node-bootstrapper) CSR: %v", err)
 			return err
 		}
-	}
-	if server {
-		if err := waitForPendingCsrs(ocConfig, "kubernetes.io/kubelet-serving"); err != nil {
-			logging.Debugf("Error waiting for second pending (system:node) CSR: %v", err)
-			return err
-		}
-		if err := ApproveNodeCSR(ocConfig, "kubernetes.io/kubelet-serving"); err != nil {
-			logging.Debugf("Error approving second pending (system:node) CSR: %v", err)
+		if err := crcerrors.RetryAfter(5*time.Minute, func() error {
+			expired, err := checkKubeletClientCertValidity(sshRunner)
+			if err != nil {
+				return err
+			}
+			if !expired {
+				return nil
+			}
+			return &crcerrors.RetriableError{Err: errors.New("kubelet client certificate still expired")}
+		}, time.Second*5); err != nil {
+			logging.Debugf("Error approving first pending (node-bootstrapper) CSR: %v", err)
 			return err
 		}
 	}
+	// API server needs to connect to kubelet for some features like logs, port forwards. This communication is backed by a cert
+	// store in /var/lib/kubelet/pki/kubelet-server-current.pem
+	// After kubelet connected to the API server, if the serving cert is expireed, kubelet asks for a new CSR.
+	// This CSR is automatically approved by the cluster-machine-approver. The k8s controller manager issues the cert and kubelet fetches it.
```

Comment:
This seems to be explaining why waitForPendingCSRs/ApproveNodeCSR was removed in the `if server` case. This should be mentioned/explained in the commit log as well.

Reply:
yes changelog is better now